### DISCUSSION TITLE
Usage of metals in the THz range

### DISCUSSION THREAD

**Discussion by Srinivas-Ganti** [NanoComp/meep#2200] [Q&A] [answered]
Hi everyone,

I am trying to simulate a metamaterial design in the THz frequency range using Meep. The details about the structure and type of measurement I am trying to simulate can be found in [this paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5560824/). Briefly, it is a transmission measurement through a periodic metal pattern on a fused silica substrate.

My question is about the material objects we import into the simulation script - when I for example import Gold (Au) from meep.materials, inspecting its "valid_frequency_range" property shows that this material definition is only valid for visible to IR wavelengths. :
```
>>> Au.valid_freq_range
FreqRange(min=0.16131113692089302, max=4.0327458966810505)
```

Does anyone have a reference/ suggestion on how we could define metals in THz frequency range ? I am interested in replicating my experimental findings in simulation. Is it possible to use Meep to do this? The typical cell volume I am dealing with here is enormous... What is the correct way to use Meep for such applications? I am posting below my attempt at setting it up but the simulation fails as the epsilon is not defined/ valid for the frequency range I am trying to simulate. 

```
import meep as mp
from meep.materials import fused_quartz, Au
import matplotlib.pyplot as plt

from HDesign import makeUnitCell 
# default unit length is 1 um
um_scale = 1.0 

resolution = 40     # pixels/um
L = 36              # outer metamaterial dimensions
g = 2               # gap size
w = 4               # track width
l = 10              # len of capacitive section
p = 50              # periodicity in x and y directions
s_l = 50            # substrate die length
s_w = 50            # substrate die width 
tmet = 0.2          # metal layer thickness
tsub = 2.0          # substrate thickness
tabs = 2            # absorber thickness
tair = 6.0          # air thickness
dpml = 2            # pml thickness

tstack = tmet + tsub + tair+ dpml+ tabs

# define cell geometry
cell_size = mp.Vector3(p,p,tstack)
# metamaterial pattern
unit_cell = makeUnitCell(L,g,w,l)
polyset = unit_cell.get_polygons(by_spec=True)
vertices = polyset[(1,0)][0]
v = [mp.Vector3(vertices[i][0],vertices[i][1]) for i in range(len(vertices))]

metam = mp.Prism(v, height = tmet, material = Au)
subs = mp.Block(material = fused_quartz, size = mp.Vector3(p,p,tsub),center = mp.Vector3(0,0,-tsub/2))

geom = [metam, subs]


boundary_layers = [mp.PML(dpml, direction=mp.Z, side=mp.High),
                   mp.Absorber(tabs, direction=mp.Z, side=mp.Low)]

lambda_min = 272.5                 # 1.1 THz or minimum source wavelength in um
lambda_max = 333.1                 # 0.9 THz or maximum source wavelength in um
fmin = 1/lambda_max
fmax = 1/lambda_min
fcen = 0.5*(fmin+fmax)
df = fmax-fmin

sources = [mp.Source(mp.GaussianSource(fcen, fwidth=df),
                        component=mp.Ex,
                        center=mp.Vector3(z=0.5*tstack-dpml-1),
                        size=mp.Vector3(p,p))]

sim = mp.Simulation(resolution = 15, cell_size = cell_size, geometry = geom, boundary_layers=boundary_layers, sources = sources, split_chunks_evenly = True, k_point=mp.Vector3())


nfreq = 50
fluxReg = mp.FluxRegion(center=mp.Vector3(0,0,-0.5*tstack+tabs+1),
                         size=mp.Vector3(p,p,0))
trans = sim.add_flux(fcen, df, nfreq, fluxReg)
vol = mp.Volume(center=mp.Vector3(0,0,-0.5*tstack+tabs+1),
                         size=mp.Vector3(p,p,0))

sim.run(until=10)
ex_data = sim.get_array(center=mp.Vector3(), size=mp.Vector3(p,p,0), component=mp.Ex)
sim.display_fluxes(trans) 
```

The geometry used in the above script is obtained by calling the function below:

```
import gdspy

def makeUnitCell(L,g,w,l,layer=0):
    a1 = (-L/2, -L/2)            #A 
    a2 = (L/2, -L/2)             #B  
    a3 = (L/2, L/2)              #C  
    a4 = (-L/2, L/2)             #D  
    a10 = (g/2+w, -l/2)          #I 
    a12 = (g/2, l/2)             #K
    a21 = (-g/2, l/2)            #T
    a23 = (-g/2-w, -l/2)         #V
    #make a unit cell 
    unit_cell = gdspy.Cell("unit", exclude_from_current=True)
    #define geometries of unit cell
    rect1 = gdspy.Rectangle(a1, (a2[0],a2[1]+w), layer = layer)
    rect2 = gdspy.Rectangle(a2, (a3[0]-w,a3[1]), layer = layer)
    rect3 = gdspy.Rectangle((a3[0],a3[1]-w),a4, layer= layer)
    rect4 = gdspy.Rectangle(a1,(a4[0]+w,a4[1]), layer= layer)
    rect5 = gdspy.Rectangle((-L/2,-w/2),(-g/2,w/2), layer= layer)
    rect6 = gdspy.Rectangle((g/2,w/2),(L/2,-w/2), layer= layer)
    rect7 = gdspy.Rectangle(a21,a23, layer= layer)
    rect8 = gdspy.Rectangle(a10,a12, layer= layer)
    #add elements to unit cell
    unit_cell.add(gdspy.boolean(rect1,[rect2,rect3,rect4,rect5,rect6,rect7,rect8],operation = "or",layer = 1))
    return unit_cell

```

**Comment by smartalecH** [comment, marked as answer]
> Does anyone have a reference/ suggestion on how we could define metals in THz frequency range ?

You can define your material over *whatever* range you want. You just need to find a Drude-Lorentz model for your material at that range (currently meep only supports Drude-Lorentz models, although we have worked on generalizing this to [arbitrarily rational functions](https://github.com/NanoComp/meep/pull/920)).

In other words, you don't have to use the material library. Although the material library is a good reference when learning *how* to define your own material model based on the parameters of a paper.

[The docs](https://meep.readthedocs.io/en/latest/Materials/) walk through this process rather thoroughly too.

**Comment by 10UAikawa** [comment]
Hi, I have a similar structure that I am trying to simulate. It is periodic in the x-y direction like your structure. I noticed when setting up the Bloch-periodic boundaries you set the k_point=mp.Vector3(). Which is just (0,0,0)? So, doesn't that make it periodic in all direction of x,y and z? I was wondering if you knew how to setup the periodic boundary conditions because this is something I am trying to do too.
